### PR TITLE
Use the new IPython completer API

### PR DESCRIPTION
**Pull request by Carreau**
Work in progress, I want to see if I can use the metadata fields for frontends to opt-in to some extra informations.

Also how backward compatible would it be to have a metadata field on the `complete_request` ? My idea is for testing would have been to have an "Accept:" like field for the kernel to know whether to compute the new completions.

### CODE REVIEW

**Comment by minrk** [comment]
> Also how backward compatible would it be to have a metadata field on the complete_request ?

Adding a field should be a minor revision, and if things work when that field is empty / missing, it's not a compatibility problem at all.

**Comment by rgbkrk** [comment]
I'm ok with introducing new fields rather than metadata for the final version of this (that can be ignored by some frontends), effectively `matches` and `enhanced_matches`.

The annoying thing I've noticed about our "metadata" fields is that people come to rely on the metadata if it gets merged in as a metadata field, and then it's no longer extraneous metadata -- it's an expected format with it's own versioning.

**Comment by Carreau** [comment]
> I'm ok with introducing new fields rather than metadata for the final version of this (that can be ignored by some frontends), effectively matches and enhanced_matches.
> 
> The annoying thing I've noticed about our "metadata" fields is that people come to rely on the metadata if it gets merged in as a metadata field, and then it's no longer extraneous metadata -- it's an expected format with it's own versioning.


Agreed, but right now adding things into the metadata field allow us to experiment without changing the wireformat and having to deal with the serialisation layer. 

Once we get the fields right it's "easy" to make an enhancement proposal to move them up one level. 

I've push a commit that send these new informations in the metadata field for now. 
We can start bikeshedding on names and necessary fields.

**Comment by rgbkrk** [comment]
Sounds great, thanks Matthias.

**Comment by minrk** [comment]
Sounds like a good plan.

File: ipykernel/ipkernel.py [hunk 1]
```diff
@@ -261,6 +270,42 @@ def do_complete(self, code, cursor_pos):
...
+            completions = list(rectify_completions(code, raw_completions))
+            
+            comps = []
+            for comp in completions:
+                comps.append(dict(
+                            start=comp.start,
+                            end=comp.end,
+                            text=comp.text,
+                            type=comp.type,
+                              ))
```
Comment (ellisonbg) [inline ipykernel/ipkernel.py]: Dedent these `))` to the level of `comps`.

File: ipykernel/ipkernel.py [hunk 2]
```diff
@@ -261,6 +270,42 @@ def do_complete(self, code, cursor_pos):
...
+
+        if completions:
+            s = completions[0].start
```
Comment (ellisonbg) [inline ipykernel/ipkernel.py]: Seems reasonable to use `completions[0]` but might there ever be a reason to not pick this one?

File: ipykernel/ipkernel.py [hunk 3]
```diff
@@ -261,6 +270,42 @@ def do_complete(self, code, cursor_pos):
...
+            matches = [c.text for c in completions]
+        else:
+            s = cursor_pos
+            e = cursor_pos
+            matches = []
+
+        return {'matches': matches,
+                'cursor_end': e,
+                'cursor_start': s,
+                'metadata': {'_experimental': comps},
```
Comment (ellisonbg) [inline ipykernel/ipkernel.py]: The only big question is how to encode this in the message data. I think this approach of putting this into the metadata makes sense. Maybe add a suffix to help a frontend to distinguish if there are multiple experimental extensions:

`_experimental_type_data`

Comment (ellisonbg) [inline ipykernel/ipkernel.py]: Some ideas:

* `_jupyter_types_prototype`
* `_jupyter_types_experimental`

File: ipykernel/ipkernel.py (hunk 2 above)
Comment (Carreau) [inline ipykernel/ipkernel.py]: No. `rectify_completion` is made to have all completions have the same beginning and end and pad these potentially ugly characters. I can re-assert that `len(set(c.start for c in completions)) == 1`, but that seem unnecessary.

Comment (ellisonbg) [inline ipykernel/ipkernel.py]: Sounds good!

**Comment by Carreau** [comment]
Rebased and comment taken into account.
This will likely also affect a bit completions in the normal notebooks, it there are bugs, please let me know.

**Comment by Carreau** [comment]
Thanks !

**Comment by rgbkrk** [comment]
😮 

🎉 🎉 🎉

**Pull request merged**